executor/src/witgen/affine_expression.rs

```diff
@@ -444,28 +421,110 @@ where
 {
     type Output = Self;
 
-    fn add(mut self, mut rhs: Self) -> Self::Output {
-        self.offset += rhs.offset;
-
-        // Combine coefficients and try to retain the clean flag.
-        if rhs.coefficients.is_empty() {
-            // All clean.
-        } else if self.coefficients.is_empty() {
-            self.coefficients = rhs.coefficients;
-            self.clean = rhs.clean;
-        } else if let [(lk, lv)] = self.coefficients[..] {
-            self.clean = !rhs.coefficients.iter().any(|(k, _)| k == &lk);
-            self.coefficients = rhs.coefficients;
-            self.coefficients.push((lk, lv));
-        } else if let [(rk, rv)] = rhs.coefficients[..] {
-            self.clean = !self.coefficients.iter().any(|(k, _)| k == &rk);
-            self.coefficients.push((rk, rv));
-        } else {
-            self.coefficients.append(&mut rhs.coefficients);
-            self.clean = false;
+    fn add(self, rhs: Self) -> Self::Output {
+        use std::cmp::Ordering;
+        use AffineExpression::*;
+        match (self, rhs) {
+            // constant + constant
+            (Constant(o1), Constant(o2)) => Constant(o1 + o2),
+            // constant + one
+            (One(coeff, o1), Constant(o2)) | (Constant(o2), One(coeff, o1)) => One(coeff, o1 + o2),
+            // constant + many
+            (
+                Many {
```

Comment:
it's weird - this is not too long to fit a single line, right? Does it always break patterns like this?

Reply:
Is there a way to use something like `many @ Many` and then use `many.coefficients` below?

Reply:
I mean another "solution" would be to not use named parameters for Many, as you do already for One.

Reply:
> Is there a way to use something like `many @ Many` and then use `many.coefficients` below?
I think not, but non-named parameters could be better indeed.